## Preprocessing

In [31]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import tensorflow as tf

#  Import and read the charity_data.csv.
import pandas as pd 
application_df = pd.read_csv("Resources/charity_data.csv")
application_df.head()

EIN                                      NAME APPLICATION_TYPE  \
0  10520599              BLUE KNIGHTS MOTORCYCLE CLUB              T10   
1  10531628    AMERICAN CHESAPEAKE CLUB CHARITABLE TR               T3   
2  10547893        ST CLOUD PROFESSIONAL FIREFIGHTERS               T5   
3  10553066            SOUTHSIDE ATHLETIC ASSOCIATION               T3   
4  10556103  GENETIC RESEARCH INSTITUTE OF THE DESERT               T3   

        AFFILIATION CLASSIFICATION      USE_CASE  ORGANIZATION  STATUS  \
0       Independent          C1000    ProductDev   Association       1   
1       Independent          C2000  Preservation  Co-operative       1   
2  CompanySponsored          C3000    ProductDev   Association       1   
3  CompanySponsored          C2000  Preservation         Trust       1   
4       Independent          C1000     Heathcare         Trust       1   

      INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  IS_SUCCESSFUL  
0              0                      N     5000              1  
1         1-9999                      N   108590              1  
2              0                      N     5000              0  
3    10000-24999                      N     6692              1  
4  100000-499999                      N   142590              1

In [32]:
# Drop the non-beneficial ID columns, 'EIN' and 'NAME'.
application_df= application_df.drop(labels=['EIN','NAME'],axis=1)

In [33]:
# Determine the number of unique values in each column.
for col in application_df.columns:
    print(col +': ' +str(len(pd.unique(application_df[col]))))

APPLICATION_TYPE: 17
AFFILIATION: 6
CLASSIFICATION: 71
USE_CASE: 5
ORGANIZATION: 4
STATUS: 2
INCOME_AMT: 9
SPECIAL_CONSIDERATIONS: 2
ASK_AMT: 8747
IS_SUCCESSFUL: 2


In [34]:
# Look at APPLICATION_TYPE value counts for binning
application_df['APPLICATION_TYPE'].value_counts()

T3     27037
T4      1542
T6      1216
T5      1173
T19     1065
T8       737
T7       725
T10      528
T9       156
T13       66
T12       27
T2        16
T25        3
T14        3
T29        2
T15        2
T17        1
Name: APPLICATION_TYPE, dtype: int64

In [35]:
# Choose a cutoff value and create a list of application types to be replaced
# use the variable name `application_types_to_replace`
other_application_types = ['T25','T14','T29','T15','T19','T2','T12','T13','T9','T17']

# Replace in dataframe
for app in other_application_types:
    application_df['APPLICATION_TYPE'] = application_df['APPLICATION_TYPE'].replace(app,"Other")

# Check to make sure binning was successful
application_df['APPLICATION_TYPE'].value_counts()

T3       27037
T4        1542
Other     1341
T6        1216
T5        1173
T8         737
T7         725
T10        528
Name: APPLICATION_TYPE, dtype: int64

In [36]:
# Look at CLASSIFICATION value counts for binning
application_df['CLASSIFICATION'].value_counts()

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
         ...  
C4120        1
C8210        1
C2561        1
C4500        1
C2150        1
Name: CLASSIFICATION, Length: 71, dtype: int64

In [37]:
# You may find it helpful to look at CLASSIFICATION value counts >1
len(application_df['CLASSIFICATION'].value_counts()[application_df['CLASSIFICATION'].value_counts()>1])

45

In [38]:
len(application_df['CLASSIFICATION'].value_counts()[application_df['CLASSIFICATION'].value_counts()>20])

20

In [39]:
other_classifications = application_df['CLASSIFICATION'].value_counts()[application_df['CLASSIFICATION'].value_counts()<20].keys()
other_classifications

Index(['C7120', 'C1500', 'C1800', 'C6000', 'C1250', 'C8200', 'C1238', 'C1278',
       'C1235', 'C1237', 'C7210', 'C2400', 'C1720', 'C4100', 'C1257', 'C1600',
       'C1260', 'C2710', 'C0', 'C3200', 'C1234', 'C1246', 'C1267', 'C1256',
       'C2190', 'C4200', 'C2600', 'C5200', 'C1370', 'C1248', 'C6100', 'C1820',
       'C1900', 'C1236', 'C3700', 'C2570', 'C1580', 'C1245', 'C2500', 'C1570',
       'C1283', 'C2380', 'C1732', 'C1728', 'C2170', 'C4120', 'C8210', 'C2561',
       'C4500', 'C2150'],
      dtype='object')

In [40]:
# Choose a cutoff value and create a list of classifications to be replaced



# Replace in dataframe
for cls in other_classifications:
    application_df['CLASSIFICATION'] = application_df['CLASSIFICATION'].replace(cls,"Other")
    
# Check to make sure binning was successful
application_df['CLASSIFICATION'].value_counts()

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
C7000      777
C1700      287
Other      207
C4000      194
C5000      116
C1270      114
C2700      104
C2800       95
C7100       75
C1300       58
C1280       50
C1230       36
C1400       34
C7200       32
C2300       32
C1240       30
C8000       20
Name: CLASSIFICATION, dtype: int64

In [41]:
# Convert categorical data to numeric with `pd.get_dummies`
application_dummies_df = pd.get_dummies(application_df)
print(application_dummies_df.columns)
application_dummies_df.head()

Index(['STATUS', 'ASK_AMT', 'IS_SUCCESSFUL', 'APPLICATION_TYPE_Other',
       'APPLICATION_TYPE_T10', 'APPLICATION_TYPE_T3', 'APPLICATION_TYPE_T4',
       'APPLICATION_TYPE_T5', 'APPLICATION_TYPE_T6', 'APPLICATION_TYPE_T7',
       'APPLICATION_TYPE_T8', 'AFFILIATION_CompanySponsored',
       'AFFILIATION_Family/Parent', 'AFFILIATION_Independent',
       'AFFILIATION_National', 'AFFILIATION_Other', 'AFFILIATION_Regional',
       'CLASSIFICATION_C1000', 'CLASSIFICATION_C1200', 'CLASSIFICATION_C1230',
       'CLASSIFICATION_C1240', 'CLASSIFICATION_C1270', 'CLASSIFICATION_C1280',
       'CLASSIFICATION_C1300', 'CLASSIFICATION_C1400', 'CLASSIFICATION_C1700',
       'CLASSIFICATION_C2000', 'CLASSIFICATION_C2100', 'CLASSIFICATION_C2300',
       'CLASSIFICATION_C2700', 'CLASSIFICATION_C2800', 'CLASSIFICATION_C3000',
       'CLASSIFICATION_C4000', 'CLASSIFICATION_C5000', 'CLASSIFICATION_C7000',
       'CLASSIFICATION_C7100', 'CLASSIFICATION_C7200', 'CLASSIFICATION_C8000',
       'CLASSIFICATION

STATUS  ASK_AMT  IS_SUCCESSFUL  APPLICATION_TYPE_Other  \
0       1     5000              1                       0   
1       1   108590              1                       0   
2       1     5000              0                       0   
3       1     6692              1                       0   
4       1   142590              1                       0   

   APPLICATION_TYPE_T10  APPLICATION_TYPE_T3  APPLICATION_TYPE_T4  \
0                     1                    0                    0   
1                     0                    1                    0   
2                     0                    0                    0   
3                     0                    1                    0   
4                     0                    1                    0   

   APPLICATION_TYPE_T5  APPLICATION_TYPE_T6  APPLICATION_TYPE_T7  ...  \
0                    0                    0                    0  ...   
1                    0                    0                    0  ...   
2                    1                    0                    0  ...   
3                    0                    0                    0  ...   
4                    0                    0                    0  ...   

   INCOME_AMT_1-9999  INCOME_AMT_10000-24999  INCOME_AMT_100000-499999  \
0                  0                       0                         0   
1                  1                       0                         0   
2                  0                       0                         0   
3                  0                       1                         0   
4                  0                       0                         1   

   INCOME_AMT_10M-50M  INCOME_AMT_1M-5M  INCOME_AMT_25000-99999  \
0                   0                 0                       0   
1                   0                 0                       0   
2                   0                 0                       0   
3                   0                 0                       0   
4                   0                 0                       0   

   INCOME_AMT_50M+  INCOME_AMT_5M-10M  SPECIAL_CONSIDERATIONS_N  \
0                0                  0                         1   
1                0                  0                         1   
2                0                  0                         1   
3                0                  0                         1   
4                0                  0                         1   

   SPECIAL_CONSIDERATIONS_Y  
0                         0  
1                         0  
2                         0  
3                         0  
4                         0  

[5 rows x 59 columns]

In [42]:

low_relational_variables = ['USE_CASE_Other','CLASSIFICATION_Other']
application_dummies_df = application_dummies_df.drop(labels=low_relational_variables, axis=1)


In [43]:
# Split our preprocessed data into our features and target arrays
X= application_dummies_df.drop('IS_SUCCESSFUL',axis=1)
y = application_dummies_df['IS_SUCCESSFUL']

# Split the preprocessed data into a training and testing dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)

In [44]:
# Create a StandardScaler instances
scaler = StandardScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)




## Compile, Train and Evaluate the Model

In [45]:

# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
input_dimensions = len(application_dummies_df.columns)-1 #all columns
low_node_count = 10
mid_node_count = 25
high_node_count = 50
model = tf.keras.models.Sequential([
  tf.keras.layers.Dense(units=high_node_count,activation='relu',input_dim=input_dimensions),
  tf.keras.layers.Dense(units=high_node_count,activation='gelu'),
  tf.keras.layers.Dense(units=mid_node_count,activation='relu'),
  tf.keras.layers.Dense(units=mid_node_count,activation='gelu'),
  tf.keras.layers.Dense(units=low_node_count,activation='relu'),
  tf.keras.layers.Dense(units=low_node_count,activation='gelu'),
  tf.keras.layers.Dense(units=1,activation='sigmoid')
])

# Check the structure of the model
model.summary()

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_7 (Dense)             (None, 50)                2850      
                                                                 
 dense_8 (Dense)             (None, 50)                2550      
                                                                 
 dense_9 (Dense)             (None, 25)                1275      
                                                                 
 dense_10 (Dense)            (None, 25)                650       
                                                                 
 dense_11 (Dense)            (None, 10)                260       
                                                                 
 dense_12 (Dense)            (None, 10)                110       
                                                                 
 dense_13 (Dense)            (None, 1)                

In [49]:
# Compile the model
model.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

model.fit(X_train_scaled, y_train, epochs=20)

Epoch 1/20
804/804 [==============================] - 2s 1ms/step - loss: 0.5376 - accuracy: 0.7387
Epoch 2/20
804/804 [==============================] - 1s 1ms/step - loss: 0.5370 - accuracy: 0.7399
Epoch 3/20
804/804 [==============================] - 1s 1ms/step - loss: 0.5370 - accuracy: 0.7395
Epoch 4/20
804/804 [==============================] - 1s 1ms/step - loss: 0.5364 - accuracy: 0.7394
Epoch 5/20
804/804 [==============================] - 1s 1ms/step - loss: 0.5364 - accuracy: 0.7393
Epoch 6/20
804/804 [==============================] - 1s 1ms/step - loss: 0.5360 - accuracy: 0.7395
Epoch 7/20
804/804 [==============================] - 1s 1ms/step - loss: 0.5357 - accuracy: 0.7392
Epoch 8/20
804/804 [==============================] - 1s 1ms/step - loss: 0.5360 - accuracy: 0.7392
Epoch 9/20
804/804 [==============================] - 1s 1ms/step - loss: 0.5357 - accuracy: 0.7395
Epoch 10/20
804/804 [==============================] - 1s 2ms/step - loss: 0.5355 - accuracy: 0.7394

In [50]:
model_loss, model_accuracy = model.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 0s - loss: 0.5527 - accuracy: 0.7298 - 274ms/epoch - 1ms/step
Loss: 0.5526723265647888, Accuracy: 0.7297959327697754


# Better training

73% accuracy is not good enough for a loan agency therefor better options must be made to get an accurate model. To do this Tensor Flow has built in optimization functions which will be used to create a better layout for the machine learning model

In [51]:
# Export our model to HDF5 file
model.save('Resources/trained_model_1.h5')

In [52]:
def create_model(hp):
    nn_model = tf.keras.models.Sequential()

    # Allow kerastuner to decide which activation function to use in hidden layers
    activation = hp.Choice('activation',['relu','gelu','softmax','tanh','sigmoid'])
    
    # Allow kerastuner to decide number of neurons in first layer
    nn_model.add(tf.keras.layers.Dense(units=hp.Int('first_units',
        min_value=20,
        max_value=50,
        step=5), activation=activation, input_dim=input_dimensions))

    # Allow kerastuner to decide number of hidden layers and neurons in hidden layers
    for i in range(hp.Int('num_layers', 3, 6)):
        nn_model.add(tf.keras.layers.Dense(units=hp.Int('units_' + str(i),
            min_value=5,
            max_value=25,
            step=5),
            activation=activation))
    
    nn_model.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

    # Compile the model
    nn_model.compile(loss="binary_crossentropy", optimizer='adam', metrics=["accuracy"])
    
    return nn_model

In [53]:
import keras_tuner as kt

tuner = kt.Hyperband(
    create_model,
    objective="val_accuracy",
    max_epochs=20,
    hyperband_iterations=2)

INFO:tensorflow:Reloading Oracle from existing project .\untitled_project\oracle.json
INFO:tensorflow:Reloading Tuner from .\untitled_project\tuner0.json


In [54]:
tuner.search(X_train_scaled,y_train,epochs=20,validation_data=(X_test_scaled,y_test))

INFO:tensorflow:Oracle triggered exit


In [55]:
best_hyper = tuner.get_best_hyperparameters(1)[0]
best_hyper.values

{'activation': 'sigmoid',
 'first_units': 35,
 'num_layers': 5,
 'units_0': 20,
 'units_1': 10,
 'units_2': 5,
 'units_3': 10,
 'units_4': 10,
 'units_5': 20,
 'tuner/epochs': 20,
 'tuner/initial_epoch': 0,
 'tuner/bracket': 0,
 'tuner/round': 0}

In [68]:
best_model = tuner.get_best_models(1)[0]
model_loss, model_accuracy = best_model.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")
best_model.summary()

268/268 - 0s - loss: 0.5562 - accuracy: 0.7341 - 296ms/epoch - 1ms/step
Loss: 0.5561676025390625, Accuracy: 0.7341107726097107
Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense (Dense)               (None, 35)                1995      
                                                                 
 dense_1 (Dense)             (None, 20)                720       
                                                                 
 dense_2 (Dense)             (None, 10)                210       
                                                                 
 dense_3 (Dense)             (None, 5)                 55        
                                                                 
 dense_4 (Dense)             (None, 10)                60        
                                                                 
 dense_5 (Dense)             (None, 10)                110   

In [57]:
best_model.save('auto_train_model.h5')

# more preprocessing

since the best accuracy is still only 73.4% other measures are going to be taken to improve the results in this case removing columns of data which may confuse the model because they are too similar in function like income groups under 100k or uncommon affiliation types.  If that is not enough a different scaling approach will be used as well.

In [58]:
application_dummies_df.columns

Index(['STATUS', 'ASK_AMT', 'IS_SUCCESSFUL', 'APPLICATION_TYPE_Other',
       'APPLICATION_TYPE_T10', 'APPLICATION_TYPE_T3', 'APPLICATION_TYPE_T4',
       'APPLICATION_TYPE_T5', 'APPLICATION_TYPE_T6', 'APPLICATION_TYPE_T7',
       'APPLICATION_TYPE_T8', 'AFFILIATION_CompanySponsored',
       'AFFILIATION_Family/Parent', 'AFFILIATION_Independent',
       'AFFILIATION_National', 'AFFILIATION_Other', 'AFFILIATION_Regional',
       'CLASSIFICATION_C1000', 'CLASSIFICATION_C1200', 'CLASSIFICATION_C1230',
       'CLASSIFICATION_C1240', 'CLASSIFICATION_C1270', 'CLASSIFICATION_C1280',
       'CLASSIFICATION_C1300', 'CLASSIFICATION_C1400', 'CLASSIFICATION_C1700',
       'CLASSIFICATION_C2000', 'CLASSIFICATION_C2100', 'CLASSIFICATION_C2300',
       'CLASSIFICATION_C2700', 'CLASSIFICATION_C2800', 'CLASSIFICATION_C3000',
       'CLASSIFICATION_C4000', 'CLASSIFICATION_C5000', 'CLASSIFICATION_C7000',
       'CLASSIFICATION_C7100', 'CLASSIFICATION_C7200', 'CLASSIFICATION_C8000',
       'USE_CASE_Commu

In [59]:
uncommon_categories = ['AFFILIATION_Other','AFFILIATION_Independent','AFFILIATION_Family/Parent',
    'INCOME_AMT_0','INCOME_AMT_1-9999','INCOME_AMT_10000-24999','INCOME_AMT_25000-99999',
    'SPECIAL_CONSIDERATIONS_N']

new_application_df = application_dummies_df.drop(uncommon_categories,axis=1)

In [60]:
# Split our preprocessed data into our features and target arrays
X= application_dummies_df.drop('IS_SUCCESSFUL',axis=1)
y = application_dummies_df['IS_SUCCESSFUL']

# Split the preprocessed data into a training and testing dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)

In [69]:
# Create a StandardScaler instances
from sklearn.preprocessing import PowerTransformer
scaler = PowerTransformer()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

c:\Users\zzpie\anaconda3\lib\site-packages\numpy\core\_methods.py:232: RuntimeWarning: overflow encountered in multiply
  x = um.multiply(x, x, out=x)
c:\Users\zzpie\anaconda3\lib\site-packages\numpy\core\_methods.py:243: RuntimeWarning: overflow encountered in reduce
  ret = umr_sum(x, axis, dtype, out, keepdims=keepdims, where=where)
c:\Users\zzpie\anaconda3\lib\site-packages\sklearn\preprocessing\_data.py:3237: RuntimeWarning: divide by zero encountered in log
  loglike = -n_samples / 2 * np.log(x_trans.var())


In [70]:
refined_model = best_model
refined_model.fit(X_train_scaled, y_train, epochs=20)

Epoch 1/20
804/804 [==============================] - 1s 1ms/step - loss: 0.5545 - accuracy: 0.7309
Epoch 2/20
804/804 [==============================] - 1s 1ms/step - loss: 0.5537 - accuracy: 0.7317
Epoch 3/20
804/804 [==============================] - 1s 932us/step - loss: 0.5531 - accuracy: 0.7312
Epoch 4/20
804/804 [==============================] - 1s 1ms/step - loss: 0.5521 - accuracy: 0.7323
Epoch 5/20
804/804 [==============================] - 1s 968us/step - loss: 0.5512 - accuracy: 0.7333
Epoch 6/20
804/804 [==============================] - 1s 1ms/step - loss: 0.5501 - accuracy: 0.7322
Epoch 7/20
804/804 [==============================] - 1s 985us/step - loss: 0.5493 - accuracy: 0.7318
Epoch 8/20
804/804 [==============================] - 1s 986us/step - loss: 0.5488 - accuracy: 0.7327
Epoch 9/20
804/804 [==============================] - 1s 1ms/step - loss: 0.5480 - accuracy: 0.7325
Epoch 10/20
804/804 [==============================] - 1s 1ms/step - loss: 0.5477 - accuracy

In [71]:
model_loss, model_accuracy = refined_model.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 0s - loss: 0.5512 - accuracy: 0.7307 - 172ms/epoch - 642us/step
Loss: 0.5512303709983826, Accuracy: 0.7307288646697998


In [72]:
refined_model.save('Resources/final_model.h5')

removing points which was thought to improve the accuracy in fact lowered it so more research into apropriate values will need to be taken to apropriately figure out how to refine the model
